ant-protocol/src/error.rs

```diff
@@ -82,3 +92,221 @@ impl From<store::Error> for Error {
+                holder: Box<NetworkAddress>,
+                key: Box<NetworkAddress>,
+            },
+            #[error("Could not Serialize/Deserialize RecordHeader to/from Record")]
+            RecordHeaderParsingFailed,
+            #[error("Could not Serialize/Deserialize Record")]
+            RecordParsingFailed,
+            #[error("The record already exists, so do not charge for it: {0:?}")]
+            RecordExists(PrettyPrintRecordKey<'static>),
+            // New variant that doesn't exist in the original Error enum
+            #[error("New error variant for testing")]
+            NewErrorVariant,
+            #[error("Unknown error variant")]
+            #[serde(other)]
+            Unknown,
+        }
+
+        // Test serialization and deserialization of ExtendedError
+        let extended_error = ExtendedError::NewErrorVariant;
+        let serialized = rmp_serde::to_vec(&extended_error).unwrap();
+
+        // Test that we can deserialize into the current Error enum
+        let deserialized: Error = rmp_serde::from_slice(&serialized).unwrap();
+        assert_eq!(deserialized, Error::Unknown);
+    }
+
+    #[test]
+    fn test_error_retro_compatibility_reduced() {
+        // Test with a struct that has a missing variant (simulating older version)
+        #[derive(Error, Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+        #[non_exhaustive]
+        enum ReducedError {
+            #[error("Chunk does not exist {0:?}")]
+            ChunkDoesNotExist(NetworkAddress),
+            #[error("Failed to deserialize hex ScratchpadAddress")]
+            ScratchpadHexDeserializeFailed,
+            // removed this variant
+            // - #[error("Failed to derive CipherText from encrypted_data")]
+            // - ScratchpadCipherTextFailed,
+            #[error("Provided cypher text is invalid")]
+            ScratchpadCipherTextInvalid,
+            #[error("There was an error getting the storecost from kademlia store")]
+            GetStoreQuoteFailed,
+            #[error("There was an error generating the payment quote")]
+            QuoteGenerationFailed,
+            #[error("Peer {holder:?} cannot find Record {key:?}")]
+            ReplicatedRecordNotFound {
+                holder: Box<NetworkAddress>,
+                key: Box<NetworkAddress>,
+            },
+            #[error("Could not Serialize/Deserialize RecordHeader to/from Record")]
+            RecordHeaderParsingFailed,
+            #[error("Could not Serialize/Deserialize Record")]
+            RecordParsingFailed,
+            #[error("The record already exists, so do not charge for it: {0:?}")]
+            RecordExists(PrettyPrintRecordKey<'static>),
+            // Missing some variants that exist in the current Error enum
+            #[error("Unknown error variant")]
+            #[serde(other)]
+            Unknown,
+        }
+
+        // Test serialization and deserialization of current Error
```

Comment:
also need to cover the other direction, i.e. serialized ReducedError to be parsed as Error, with both simple and complex entry.